### PR TITLE
Add FromParts documentation

### PR DESCRIPTION
**Pull request by hlbarber**
## Motivation and Context

The `FromParts` trait is an important feature that needs to be documented.

## Description

Add documentation concerning the `FromParts` trait.

## Notes

This will have to be updated if we decide to introduce something like a `IntoParts` trait.

### CODE REVIEW

File: design/src/server/from-parts.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+This allows the service builder to accept the following handler
+
+```rust
+async fn handler(input: ModelInput, extension: Extension<SomeStruct>) -> ModelOutput {
+    /* ... */
+}
+```
+
+where `ModelInput` and `ModelOutput` are specified by the Smithy Operation and `SomeStruct` is a struct which has been inserted, by middleware, into the [`http::Request::extensions`](https://docs.rs/http/latest/http/request/struct.Request.html#method.extensions). Up to 32 structures implementing `FromParts` can be provided to the handler with the constraint that they _must_ be provided _after_ the `ModelInput`.
```
Comment (jjant) [inline design/src/server/from-parts.md:69]: If multiple `impl FromParts` arguments fail to be extracted, do all the errors get represented in the response?

File: design/src/server/from-parts.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+{
+    type Rejection = MissingExtension;
+
+    fn from_parts(parts: &mut http::request::Parts) -> Result<Self, Self::Rejection> {
+        parts.extensions.remove::<T>().map(Extension).ok_or(MissingExtension)
+    }
+}
+```
+
+This allows the service builder to accept the following handler
```
[the remark this answers is not in the document]
Comment (hlbarber) [reply design/src/server/from-parts.md:61]: Yeah we're in a catch 22 at moment without https://github.com/awslabs/smithy-rs/pull/1886 a lot of this stuff is `#[doc(hidden)]` and hence we have no docs to link to.

Here is an issue tracking this problem https://github.com/awslabs/smithy-rs/issues/1931

File: design/src/server/from-parts.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (hlbarber) [reply design/src/server/from-parts.md:69]:
```rust
impl<Protocol, T> FromParts<Protocol> for Extension<T>
where
    T: Clone + Send + Sync + 'static,
{
    type Rejection = MissingExtension;

    fn from_parts(parts: &mut http::request::Parts) -> Result<Self, Self::Rejection> {
        parts.extensions.remove::<T>().map(Extension).ok_or(MissingExtension)
    }
}
```

so the second one will fail.

Comment (hlbarber) [reply design/src/server/from-parts.md:69]: Nope, only the first one will be returned via its `IntoResponse` impl.

File: design/src/server/from-parts.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (jjant) [reply design/src/server/from-parts.md:61]: Thank you for creating the issue 😄.

File: design/src/server/from-parts.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (jjant) [reply design/src/server/from-parts.md]: My bad, I missed the `remove()`. Thanks!

**Review by jjant — APPROVED** [review]

**Pull request merged**